optuna/_gp/acqf.py

```diff
@@ -45,52 +45,44 @@ def standard_logei(z: torch.Tensor) -> torch.Tensor:
     return vals
 
 
-def logei(mean: torch.Tensor, var: torch.Tensor, f0: torch.Tensor) -> torch.Tensor:
+def logei(mean: torch.Tensor, var: torch.Tensor, f0: float) -> torch.Tensor:
     # Return E_{y ~ N(mean, var)}[max(0, y-f0)]
     sigma = torch.sqrt(var)
     st_val = standard_logei((mean - f0) / sigma)
     val = torch.log(sigma) + st_val
     return val
 
 
-def eval_logei(
-    kernel_params: KernelParamsTensor,
-    X: torch.Tensor,
-    is_categorical: torch.Tensor,
-    cov_Y_Y_inv: torch.Tensor,
-    cov_Y_Y_inv_Y: torch.Tensor,
-    max_Y: torch.Tensor,
-    x: torch.Tensor,
-    # Additional noise to prevent numerical instability.
-    # Usually this is set to a very small value.
-    stabilizing_noise: float,
-) -> torch.Tensor:
-    cov_fx_fX = kernel(is_categorical, kernel_params, x[..., None, :], X)[..., 0, :]
-    cov_fx_fx = kernel_at_zero_distance(kernel_params)
-    (mean, var) = posterior(cov_Y_Y_inv, cov_Y_Y_inv_Y, cov_fx_fX, cov_fx_fx)
-    val = logei(mean, var + stabilizing_noise, max_Y)
-
-    return val
+def ucb(mean: torch.Tensor, var: torch.Tensor, beta: float) -> torch.Tensor:
+    return mean + torch.sqrt(beta * var)
 
 
 # TODO(contramundum53): consider abstraction for acquisition functions.
+class AcquisitionFunctionType(IntEnum):
+    LOG_EI = 0
+    UCB = 1
+
+
 @dataclass(frozen=True)
 class AcquisitionFunctionParams:
```

Comment:
Can we do something like this?

```python
@dataclass(frozen=True)
class BaseAcquisitionFunction:
    kernel_params: KernelParamsTensor
    X: np.ndarray
    search_space: SearchSpace
    cov_Y_Y_inv: np.ndarray
    cov_Y_Y_inv_Y: np.ndarray
    acqf_stabilizing_noise: float

    def _posterior(self, x: torch.Tensor) -> tuple[torch.Tensor, torch.Tensor]:
        is_categorical = torch.from_numpy(self.search_space.scale_types == ScaleType.CATEGORICAL)
        cov_fx_fX = torch.from_numpy(kernel(is_categorical, self.kernel_params, x[..., None, :], self.X)[..., 0, :])
        cov_fx_fx = torch.from_numpy(kernel_at_zero_distance(self.kernel_params))
        cov_Y_Y_inv = torch.from_numpy(self.cov_Y_Y_inv)
        cov_Y_Y_inv_Y = torch.from_numpy(self.cov_Y_Y_inv_Y)       

        # mean = cov_fx_fX @ inv(cov_fX_fX + noise * I) @ Y
        # var = cov_fx_fx - cov_fx_fX @ inv(cov_fX_fX + noise * I) @ cov_fx_fX.T

        mean = cov_fx_fX @ cov_Y_Y_inv_Y  # [batch]
        var = cov_fx_fx - (cov_fx_fX * (cov_fx_fX @ cov_Y_Y_inv)).sum(dim=-1)  # [batch]
        # We need to clamp the variance to avoid negative values due to numerical errors.
        return (mean, torch.clamp(var, min=0.0))

    def compute_with_no_grad(self, x: np.ndarray) -> np.ndarray:
        with torch.no_grad():
            return self.compute(torch.from_numpy(x)).detach().numpy()

    @abstractmethod
    def compute(self, x: torch.Tensor) -> torch.Tensor:
        raise NotImplementedError


@dataclass(frozen=True)
class UCB(BaseAcquisitionFunction):
    beta: float

    def compute(self, x: torch.Tensor) -> torch.Tensor:
        mean, var = self._posterior(x)
        return mean + torch.sqrt(self.beta * var)

@dataclass(frozen=True)
class LogEI(BaseAcquisitionFunction):
    max_Y: float

    def compute(self, x: torch.Tensor) -> torch.Tensor:
        # Return E_{y ~ N(mean, var)}[max(0, y-max_Y]
        mean, var = self._posterior(x)
        sigma = torch.sqrt(var)
        st_val = standard_logei((mean - self.max_Y) / sigma)
        return torch.log(sigma) + st_val
```

Reply:
`numba` cannot support polymorphism like this.